### PR TITLE
Add random errors in AdService

### PR DESCRIPTION
**Pull request by jack-berg**
Generate random errors that exercise error handling login in the AdService.  Currently, the error handling logic which adds an event and sets the span status isn't reachable, as far as I can tell.

### CODE REVIEW

**Review by mviitane — COMMENTED** [review]
What would be the reason for creating this error in the normal flow? Should this logic be behind a feature flag (disabled by default)? I’d expect the demo app to run without errors by default unless specifically breaking it.

**Comment by julianocosta89** [comment]
I agree with @mviitane, if we want to add errors to `AdService` we should add it behind a FeatureFlag as we have for other services.

Don't get me wrong @jack-berg I think the PR is valid and it adds a scenario to the demo.
We just need to make it possible to enable/disable the errors from the `FeatureFlagService`.

You can either add this to the current PR, or if you prefer, we can close this one and you can send a new PR with the implementation.

Here is an example of how that is handled in the `ProductCatalogService`: https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/productcatalogservice/main.go#L207

**Comment by julianocosta89** [comment]
> Adding this as a feature flag sounds reasonable to me. I tried to make progress on this today but was [having some trouble](https://github.com/open-telemetry/opentelemetry-demo/issues/696) running the featureflagservice locally, so I'm blocked for the moment.

@jack-berg the featureflagservice was fixed yesterday 🥳 
You should be able to build it locally now.

File: src/featureflagservice/priv/repo/migrations/20220524172636_create_featureflags.exs [hunk 1]
```diff
@@ -25,6 +25,11 @@ defmodule Featureflagservice.Repo.Migrations.CreateFeatureflags do
       name: "recommendationCache",
       description: "Cache recommendations",
       enabled: false})
+
+    repo().insert(%Featureflagservice.FeatureFlags.FeatureFlag{
+      name: "adServiceFailure",
+      description: "Fail ad service requests sporadically",
+      enabled: false})
```
Comment (jack-berg) [inline src/featureflagservice/priv/repo/migrations/20220524172636_create_featureflags.exs:32]: Copy and pasted. Do I need to do anything else to ensure that migration runs and applies this change?

File: docker-compose.yml [hunk 2]
```diff
@@ -74,6 +74,7 @@ services:
       - "${AD_SERVICE_PORT}"
     environment:
       - AD_SERVICE_PORT
+      - FEATURE_FLAG_GRPC_SERVICE_ADDR
```
Comment (jack-berg) [inline docker-compose.yml:77]: Need to make a corresponding update to the [helm charts](https://github.com/open-telemetry/opentelemetry-helm-charts).

File: src/featureflagservice/priv/repo/migrations/20220524172636_create_featureflags.exs (hunk 1 above)
Comment (julianocosta89) [reply src/featureflagservice/priv/repo/migrations/20220524172636_create_featureflags.exs:32]: Great question,  I think @tsloughter would be able to help us out.
If I'm not mistaken @puckpuck also added a FeatureFlag previously.

Comment (mviitane) [reply src/featureflagservice/priv/repo/migrations/20220524172636_create_featureflags.exs:32]: @jack-berg, see the last 5 lines of this file. I think the new flag should be added also there.

File: src/adservice/src/main/java/hipstershop/AdService.java [hunk 3]
```diff
@@ -156,7 +182,15 @@ public void getAds(AdRequest req, StreamObserver<AdResponse> responseObserver) {
...
         span.setAttribute("app.ads.ad_response_type", adResponseType.name());
 
-        adRequestsCounter.add(1, Attributes.of(adRequestTypeKey, adRequestType.name(), adResponseTypeKey, adResponseType.name()));
+        adRequestsCounter.add(
+            1,
+            Attributes.of(
+                adRequestTypeKey, adRequestType.name(), adResponseTypeKey, adResponseType.name()));
+
+        if (checkAdFailure()) {
+          logger.warn(ADSERVICE_FAIL_FEATURE_FALG + " fail feature flag enabled");
```
Comment (mviitane) [inline src/adservice/src/main/java/hipstershop/AdService.java:191]: FALG->FLAG

File: src/adservice/src/main/java/hipstershop/AdService.java [hunk 4]
```diff
@@ -105,15 +121,25 @@ private void stop() {
...
 
   private enum AdResponseType {
-    TARGETED, RANDOM
+    TARGETED,
+    RANDOM
   }
 
   private static class AdServiceImpl extends hipstershop.AdServiceGrpc.AdServiceImplBase {
 
+    private static final String ADSERVICE_FAIL_FEATURE_FALG = "adServiceFailure";
```
Comment (mviitane) [inline src/adservice/src/main/java/hipstershop/AdService.java:135]: FALG->FLAG

File: src/adservice/src/main/java/hipstershop/AdService.java [hunk 5]
```diff
@@ -169,6 +203,20 @@ public void getAds(AdRequest req, StreamObserver<AdResponse> responseObserver) {
...
+    boolean checkAdFailure() {
+      // Flip a coin and fail 1/10th of the time if feature flag is enabled
+      if (random.nextInt(10) != 1) {
+        return false;
+      }
+
+      GetFlagResponse response =
+          featureFlagServiceStub.getFlag(
+              hipstershop.Demo.GetFlagRequest.newBuilder()
+                  .setName(ADSERVICE_FAIL_FEATURE_FALG)
```
Comment (mviitane) [inline src/adservice/src/main/java/hipstershop/AdService.java:216]: FALG->FLAG

**Review by mviitane — COMMENTED** [review]
Running fine on my docker environment. Added few minor comments.

File: src/adservice/settings.gradle [hunk 6]
```diff
@@ -1 +1 @@
-rootProject.name = 'hipstershop'
+rootProject.name = 'opentelemetry-demo-ad-service'
```
Comment (jack-berg) [inline src/adservice/settings.gradle:1]: Giving the adservice module a more appropriate name. This property is used by IDEs as the name of the project, so this makes it easier to find in a list. `Dockerfile` needed a small update as well.

File: src/featureflagservice/priv/repo/migrations/20220524172636_create_featureflags.exs (hunk 1 above)
Comment (jack-berg) [reply src/featureflagservice/priv/repo/migrations/20220524172636_create_featureflags.exs:32]: Thanks!

**Review by mviitane — APPROVED** [review]
Changelog conflict needs to be resolved. Otherwise, LGTM.

**Review by austinlparker — APPROVED** [review]

**Review by julianocosta89 — APPROVED** [review]
LGTM

**Pull request merged**